django/core/handlers/asgi.py

```diff
@@ -0,0 +1,325 @@
+            )
+            if 'charset' in self.content_params:
+                try:
+                    codecs.lookup(self.content_params['charset'])
+                except LookupError:
+                    pass
+                else:
+                    self.encoding = self.content_params['charset']
+        else:
+            self.content_type, self.content_params = '', {}
+        # Pull out content length info
+        if self.META.get('CONTENT_LENGTH', None):
+            try:
+                self._content_length = int(self.META['CONTENT_LENGTH'])
+            except (ValueError, TypeError):
+                pass
+        # Directly assign the body file to be our stream
+        self._stream = body_file
+        # Other bits
+        self.resolver_match = None
+
+    @cached_property
+    def GET(self):
+        return http.QueryDict(self.META['QUERY_STRING'])
+
+    def _get_scheme(self):
+        return self.scope.get('scheme') or super()._get_scheme()
+
+    def _get_post(self):
+        if not hasattr(self, '_post'):
+            self._read_started = False
+            self._load_post_and_files()
+        return self._post
+
+    def _set_post(self, post):
+        self._post = post
+
+    def _get_files(self):
+        if not hasattr(self, '_files'):
+            self._read_started = False
+            self._load_post_and_files()
+        return self._files
+
+    POST = property(_get_post, _set_post)
+    FILES = property(_get_files)
+
+    @cached_property
+    def COOKIES(self):
+        return http.parse_cookie(self.META.get('HTTP_COOKIE', ''))
+
+
+class ASGIHandler(base.BaseHandler):
+    """
+    Handler for ASGI requests.
+    """
+    request_class = ASGIRequest
+    # Size to chunk response bodies into for multiple response messages.
+    chunk_size = 2 ** 16
+
+    def __init__(self):
+        super(ASGIHandler, self).__init__()
+        self.load_middleware()
+
+    async def __call__(self, scope, receive, send):
+        """
+        Async entrypoint - parses the request and hands off to get_response.
+        """
+        # Only serve HTTP connections
+        # (we should allow some way to override this)
+        if scope['type'] != 'http':
+            raise ValueError(
+                'Django can only handle ASGI/HTTP connections, not %s'
+                % scope['type']
+            )
+        # Receive the HTTP request body as a stream object
+        try:
+            body_file = await self.read_body(receive)
+        except RequestAborted:
+            return
+        # Request is complete and can be served.
+        set_script_prefix(self.get_script_prefix(scope))
+        await sync_to_async(signals.request_started.send)(sender=self.__class__, scope=scope)
+        # Get the request and check for basic issues
+        request, error_response = self.create_request(scope, body_file)
+        if request is None:
+            await self.send_response(error_response, send)
```

Comment:
I'm not sure why, but I would use `if error_response` instead of `if request is None`.

Reply:
I would prefer this, as, in my mind, "request is None" is the contract that says it's bad.